### PR TITLE
Reconnect

### PR DESCRIPTION
**Pull request by goFrendiAsgard**
In this PR, I try to add reconnect ability.

Notable changes:

- I delete `service_amqp.js` and use `Rabbit.js` instead. It is necessary since connection and channels should be tightly coupled (once the connection re-created, the channels should be re-created as well)
- In this version, when a `pub` or `pubQueue` component created, queues & exchanges are not automatically created. Instead, they will be created when they are really needed (right just before publishing message).

To test, please install docker, modify `test/test.env` and perform `source test/test.env && npm test`. If the test is failed, please try it again. I still have no clue why it sometime doesn't work. Probably because publisher run first while subscriber is still initiating. Most likely it has something to do with my refactoring decision (change generator into async-await)

To see how `Rabbit.js` works, plese run `source test/test.env && node labs/rabbit-reconnect-test.js`

In my machine, the unit-test is pass:
screenshot from 2019-03-02 17-10-41

### CODE REVIEW

File: lib/service_pub_rabbit.js [hunk 1]
```diff
@@ -21,50 +22,32 @@ class ServicePubRabbit extends Component.mixin(AsyncEmitter) {
...
-        }
-    }
-
-    *createPublisher(event) {
-        let channel = yield this._connection.createChannel();
-        let exchangeName = this._namespace + "." + this.SERVICE_NAME + "." + event;
-        this._channels[event] = channel;
-
-        yield channel.assertExchange(exchangeName, "fanout", { durable: true });
+        const rabbit = new Rabbit({
```
Comment (reyhansofian) [inline lib/service_pub_rabbit.js]: with this approach, we will instantiate `Rabbit` every time we publish message?

Comment (reyhansofian) [reply lib/service_pub_rabbit.js]: should it better to instantiate `Rabbit` class first on constructor before we use it on this method?

Comment (reyhansofian) [reply lib/service_pub_rabbit.js]: or maybe make it singleton?

**Comment by reyhansofian** [comment]
@goFrendiAsgard since this PR use `async` feature, does this mean we need to add minimum NodeJS version? otherwise, we need to handle backward compatibility 😅

**Comment by goFrendiAsgard** [comment]
Gonna make a new PR with singleton model

**Pull request closed without merging**